test/Dialect/FIRRTL/layers-errors.mlir

```diff
@@ -0,0 +1,92 @@
+// RUN: circt-opt %s -split-input-file -verify-diagnostics
+
+firrtl.circuit "CastAwayColor" {
+  firrtl.layer @A bind {}
+  firrtl.module @CastAwayColor() {
+    %w = firrtl.wire : !firrtl.probe<uint<1>, @A>
+    // expected-error @below {{cannot discard layer requirements of input ref}}
+    // expected-note  @below {{discarding layer requirements: @A}}
+    %s = firrtl.ref.cast %w : (!firrtl.probe<uint<1>, @A>) -> !firrtl.probe<uint<1>>
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @UncoloredDefineUnderLayerblock(out %o : !firrtl.probe<uint<1>>) {
+    %c = firrtl.constant 0 : !firrtl.uint<1>
+    %r = firrtl.ref.send %c : !firrtl.uint<1>
+    firrtl.layerblock @A {
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    firrtl.layerblock @A {
+      %c = firrtl.constant 0  : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %r : !firrtl.probe<uint<1>>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.layer @B bind {}
+  firrtl.module @Foo(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @B {
+      %c = firrtl.constant 0 : !firrtl.uint<1>
+      %r = firrtl.ref.send %c : !firrtl.uint<1>
+      %s = firrtl.ref.cast %r : (!firrtl.probe<uint<1>>) -> !firrtl.probe<uint<1>, @B>
+      firrtl.ref.define %o, %s : !firrtl.probe<uint<1>, @B>
+    }
+  }
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>, @B>) {
+    firrtl.layerblock @A {
+      %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>, @B>)
+      // expected-error @below {{'firrtl.ref.define' op has more layer requirements than destination}}
+      // expected-note  @below {{additional layers required: @A}}
+      firrtl.ref.define %o, %foo_o : !firrtl.probe<uint<1>, @B>
+    }
+  }
+}
+
+// -----
+
+firrtl.circuit "Top" {
+  firrtl.layer @A bind {}
+  firrtl.extmodule @Foo(out o : !firrtl.probe<uint<1>>) attributes {layer = [@A]}
+  firrtl.module @Top(out %o : !firrtl.probe<uint<1>>) {
+    %foo_o = firrtl.instance foo @Foo(out o : !firrtl.probe<uint<1>>)
+    firrtl.ref.define %o, %foo_o : !firrtl.probe<uint<1>>
+  }
+}
```

Comment:
Nit: Would this be better placed in `layers.mlir` instead of `layers-errors.mlir`?

Reply:
Removed. this shouldn't have been committed.